docs/rfcs/20220119_raft_replication_simplification.md

```diff
@@ -0,0 +1,162 @@
+This solution may be a problem since user has an impression that data are 
+replicated while the data are actually stored only on the leader. 
+
+In the proposed solution we are going to handle back pressure in the same way 
+for both consistency level. This way there will be a guarantee that followers 
+can always keep up with the leader.
+
+
+### Memory usage
+
+Memory usage control will be achieved in the same way as it is done right now 
+i.e. we will not finish first stage of replication when `replicate_batcher` 
+memory semaphore is exhausted. This way we will prevent Kafka layer from 
+buffering indefinite number of batches in pending replicate requests. Replicate 
+batcher memory units are released after the request is handed of to the 
+socket so that all the user space buffer can be freed. 
+In future we may consider per shard semaphore controlling the total amount 
+of memory used by all replicate batchers.
+
+### Recovery STM coordination - finishing recovery
+
+`recovery_stm` is started every time the leader identifies the follower as 
+being behind. Since recovery stm is currently used for both recovering the 
+follower (when it is actually behind) and delivering follower data replicated 
+with `leader_ack` consistency the `recovery_stm` implementation contains some 
+optimization to minimize the `leader_ack` latency. 
+
+There are two main optimizations:
+- not stopping recovery when last replicate request consistency level is `leader_ack`
+- triggering next dispatch loop immediately after leader disk append
+
+Those optimizations make handling of recovering follower more complex in 
+other scenarios f.e. `leadership_transfer`. 
+
+Since there is no coordination between the two paths that the follower append 
+entries requests are sent from (i.e. `replicate_stm` and `recovery_stm`) those 
+request may be reordered and cause log truncations and redeliveries which lead 
+to increased latency.
+
+In the proposed solution since we are not longer going to relay on the 
+recovery STM to deliver messages to follower in normal conditions the complicated
+handling of recovery finishing logic will no longer be needed. 
+When follower is up to date with the leader `recovery_stm` is simply not running.
+
+The only place that will require coordination is the time when follower 
+is not longer behind the leader but `replicate_stm` skips dispatching requests 
+to the follower as it is still recovering. This may lead to the situation in 
+which next leader append will happen before `recovery_stm` received follower 
+response for last request i.e. it can not finish recovery as log end already 
+advanced. This case may be handled by coordinating dispatching follower 
+requests from `recovery_stm` and `replicate_stm`. 
+We may introduce last sent offset to 
+`raft::follower_index_metadata` to check if recovery should finish 
+instead of the check being based only on the follower response.
```

Comment:
```suggestion
The only situation that will require coordination is when the follower 
is no longer behind the leader, but `replicate_stm` skips dispatching requests 
to the follower as it is still recovering. This may lead to a situation in 
which the next leader append will happen before the `recovery_stm` received the follower 
response for the last request, i.e. it cannot finish recovery because the log end already 
advanced. This case may be handled by coordinating dispatching follower 
requests from `recovery_stm` and `replicate_stm`. 
We can introduce the last sent offset to 
`raft::follower_index_metadata` to check whether recovery should finish 
instead of the check being based only on the follower response.
```